pkg/iam/iam.go

```diff
@@ -142,3 +163,112 @@ func (c *IAMClient) OrganizationIAM(ctx context.Context, organizationID string)
 
 	return m, nil
 }
+
+// RemoveProjectIAM removes the given IAM policy membership.
+func (c *IAMClient) RemoveProjectIAM(ctx context.Context, iamMember *assetinventory.AssetIAM) error {
+	// The Golang API does not support IAM removal on projects yet.
+	args := []string{
+		"projects", "remove-iam-policy-binding",
+		iamMember.ResourceID,
+		fmt.Sprintf("--member=%s", iamMember.Member),
+		fmt.Sprintf("--role=%s", iamMember.Role),
+	}
+	if iamMember.Condition != nil {
+		args = append(args, fmt.Sprintf("--condition=%s", conditionString(*iamMember.Condition)))
+	}
+	if _, err := c.gcloud(ctx, args); err != nil {
+		return fmt.Errorf("failed to remove iam policy for project %s: %w", iamMember.ResourceID, err)
+	}
+
+	return nil
+}
+
+// RemoveFolderIAM removes the given IAM policy membership.
+func (c *IAMClient) RemoveFolderIAM(ctx context.Context, iamMember *assetinventory.AssetIAM) error {
+	// The Golang API does not support IAM removal on folders yet.
+	args := []string{
+		"resource-manager", "folders", "remove-iam-policy-binding",
+		iamMember.ResourceID,
+		fmt.Sprintf("--member=%s", iamMember.Member),
+		fmt.Sprintf("--role=%s", iamMember.Role),
+	}
+	if iamMember.Condition != nil {
+		args = append(args, fmt.Sprintf("--condition=%s", conditionString(*iamMember.Condition)))
+	}
+	if _, err := c.gcloud(ctx, args); err != nil {
+		return fmt.Errorf("failed to remove iam policy for folder %s: %w", iamMember.ResourceID, err)
+	}
+
+	return nil
+}
+
+// RemoveOrganizationIAM removes the given IAM policy membership.
+func (c *IAMClient) RemoveOrganizationIAM(ctx context.Context, iamMember *assetinventory.AssetIAM) error {
+	// The Golang API does not support IAM removal on organizations yet.
+	args := []string{
+		"organizations", "remove-iam-policy-binding",
+		iamMember.ResourceID,
+		fmt.Sprintf("--member=%s", iamMember.Member),
+		fmt.Sprintf("--role=%s", iamMember.Role),
+	}
+	if iamMember.Condition != nil {
+		args = append(args, fmt.Sprintf("--condition=%s", conditionString(*iamMember.Condition)))
+	}
+	if _, err := c.gcloud(ctx, args); err != nil {
+		return fmt.Errorf("failed to remove iam policy for organization %s: %w", iamMember.ResourceID, err)
+	}
+
+	return nil
+}
+
+// gcloud runs the gcloud command.
+func (c *IAMClient) gcloud(ctx context.Context, args []string) (int, error) {
```

Comment:
As a follow up, can we see about using the asset inventory api instead of gcloud? The only time we want to wrap gcloud is if there is some custom logic that the CLI is doing (for example gcloud run deploy from source combines all sorts of apis together)

Reply:
There is no API in golang for removing IAM AFAIK. Chen found one for setting IAM policy but that feels like not a good pattern as it assumes that you can set policy for all of your org and calls into question who owns what (instead guardian terraform should own policy/bindings).

Reply:
I believe that's what gcloud does behind the scene. It iterate through the bindings in the policy, remove the matching ones and then set it back. It has some etag as optimistic lock.

Reply:
If there is a good reason to use gcloud, e.g. arguments above, I am fine with that but its a dependency we now require to use guardian. Let's document why if we can.

Reply:
Per Seths comment, lets switch to using the API mirroring the sample he provided.

Reply:
Done